### PR TITLE
#62: Implement splitPath

### LINKED ISSUE #62: Implement `splitPath`
## Background
Operating with field name containing the whole path (parent fields)

## Feature
Split the path into its components.

## Example
`"parent.path.my_field"` -> `("parent", "path", "my_field")`

### PR DESCRIPTION
**Pull request by benedeki**
* `splitPath` function implemented and used
* Updated UTs

Closes #62

### CODE REVIEW

File: spark-commons/src/main/scala/za/co/absa/spark/commons/utils/SchemaUtils.scala [hunk 1]
```diff
@@ -93,4 +93,22 @@ object SchemaUtils {
...
 
+  /**
+   * Separates the field name components of a fully qualified column name as they hierachy goes from root down to the
+   * deepest one.
+   * Trailing '.' is ignored, leading one not.
+   *
+   * @param columnName  A fully qualified column name
+   * @return Each level field name in sequence how they go from root to the lowest one
+   */
+  def splitPath(columnName: String): List[String] = {
```
Comment (benedeki) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/utils/SchemaUtils.scala:104]: I thought returning `Vector[String]` here, as a generally more efficient data type than `List[String]`, but couple of places where this function is used expect `List` and so far no place that would benefit from `Vector` has been spotted.

**Review by miroslavpojer — DISMISSED** [review]
No review comments.

File: spark-commons/src/main/scala/za/co/absa/spark/commons/utils/SchemaUtils.scala [hunk 2]
```diff
@@ -93,4 +93,22 @@ object SchemaUtils {
     }
   }
 
+  /**
+   * Separates the field name components of a fully qualified column name as they hierachy goes from root down to the
+   * deepest one.
```
Comment (dk1844) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/utils/SchemaUtils.scala:98]:
```suggestion
   * Separates the field name components of a fully qualified column name as they hierachy goes from root down to the
   * deepest one. Example: `com.my.package.xyz` -> List(com, my, package, xyz)
```

File: spark-commons/src/test/scala/za/co/absa/spark/commons/utils/SchemaUtilsTest.scala [hunk 3]
```diff
@@ -36,4 +36,15 @@ class SchemaUtilsTest extends AnyFunSuite with Matchers {
     assertResult(getParentPath("a.bcd"))("a")
     assertResult(getParentPath(""))("")
   }
+
+  test("Test splitPath") {
+    assertResult(List("a", "b", "c", "d", "e"))(splitPath("a.b.c.d.e"))
+    assertResult(List("a"))(splitPath("a"))
+    assertResult(List("a", "bcd"))(splitPath("a.bcd"))
+    assertResult(List("a", "bcd"))(splitPath("a.bcd."))
+    assertResult(List("", "a", "bcd"))(splitPath(".a.bcd"))
```
Comment (dk1844) [inline spark-commons/src/test/scala/za/co/absa/spark/commons/utils/SchemaUtilsTest.scala:45]:
```suggestion
    assertResult(List("", "a", "bcd"))(splitPath(".a.bcd"))
    assertResult(List("", "", "", "a", "bcd"))(splitPath("...a.bcd"))
    assertResult(List("a", "bcd"))(splitPath("a.bcd."))
    assertResult(List("", "", "a", "bcd"))(splitPath("a.bcd...")) // is this expected behavior?
```
worth to demonstrate the trailing `.` corner cases, too?

**Review by dk1844 — COMMENTED** [review]
looks good overall, I am just not user about the intended behavior for possible multiple  leading/trailing `.`s.

File: spark-commons/src/main/scala/za/co/absa/spark/commons/utils/SchemaUtils.scala (hunk 1 above)
Comment (jakipatryk) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/utils/SchemaUtils.scala:104]: Tbh I expect path being too short to need such optimisations. List is fine to me.

File: README.md [hunk 4]
```diff
@@ -67,6 +67,14 @@ _SchemaUtils_ provides methods for working with schemas, its comparison and alig
       SchemaUtils.appendPath(path, fieldName)
     ```
 
+5. Separates the field name components of a fully qualified column name as they hierachy goes from root down to the
+deepest one.
+
+    ```scala
+      SchemaUtils.splitPath(path, fieldName)
```
Comment (jakipatryk) [inline README.md:74]: Shouldn't be without `fieldName`?

Comment (benedeki) [reply README.md:74]: Sure, copy paste error. 😊

File: spark-commons/src/test/scala/za/co/absa/spark/commons/utils/SchemaUtilsTest.scala (hunk 3 above)
Comment (benedeki) [reply spark-commons/src/test/scala/za/co/absa/spark/commons/utils/SchemaUtilsTest.scala:45]: I was wondering about the last case too (multiple dots and the end). Not sure which is the most logical/expected behavior. 🤔 Any opinions?

Comment (miroslavpojer) [reply spark-commons/src/test/scala/za/co/absa/spark/commons/utils/SchemaUtilsTest.scala:45]: Are we providing only the service of "path splitting"? With available information we cannot decide what is valid or invalid input: "...ab.cd" or "ab.cd...." or "ab...cd".

File: spark-commons/src/main/scala/za/co/absa/spark/commons/utils/SchemaUtils.scala [hunk 5]
```diff
@@ -89,7 +89,7 @@ object SchemaUtils {
    * @param path      The dot-separated existing path
    * @param fieldName Name of the field to be appended to the path
    * @return The path with the new field appended or the field itself if path is empty
-   */
+   */()
```
Comment (Zejnilovic) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/utils/SchemaUtils.scala:92]: What is this?

Comment (benedeki) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/utils/SchemaUtils.scala:92]: No idea. Will remove.

File: spark-commons/src/test/scala/za/co/absa/spark/commons/utils/SchemaUtilsTest.scala (hunk 3 above)
Comment (benedeki) [reply spark-commons/src/test/scala/za/co/absa/spark/commons/utils/SchemaUtilsTest.scala:54]: I was thinking about possible approaches. Let's vote (with emoticons):
* Keep as it is 👍 
* Throw exception if there's an empty field name 🚀 
* Remove all empty field names from result list 🎉 

I still think that one closing dot (`.`) could still be accepted.

Comment (dk1844) [reply spark-commons/src/test/scala/za/co/absa/spark/commons/utils/SchemaUtilsTest.scala:54]: My vote would be
```scala
def splitPath(path: String, keepEmptyFields: Boolean = true)
```
Had I had to choose from the above, I'd go 🎉 .

File: spark-commons/src/main/scala/za/co/absa/spark/commons/utils/SchemaUtils.scala [hunk 6]
```diff
@@ -100,6 +100,44 @@ object SchemaUtils {
     }
   }
 
+
+  /**
+   * Separates the field name components of a fully qualified column name as they hierarchy goes from root down to the
```
Comment (benedeki) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/utils/SchemaUtils.scala:105]:
```suggestion
   * Separates the field name components of a fully qualified column name as their hierarchy goes from root down to the
```

[the remark this answers is not in the document]
Comment (benedeki) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/utils/SchemaUtils.scala:105]: No idea about the code duplication. 🤷

File: spark-commons/src/test/scala/za/co/absa/spark/commons/utils/SchemaUtilsTest.scala (hunk 3 above)
Comment (miroslavpojer) [reply spark-commons/src/test/scala/za/co/absa/spark/commons/utils/SchemaUtilsTest.scala:54]: Tests Passed.

**Review by jakipatryk — APPROVED** [review]
LGTM

File: spark-commons/src/main/scala/za/co/absa/spark/commons/utils/SchemaUtils.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (lsulak) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/utils/SchemaUtils.scala]: I though about something like this: https://stackoverflow.com/questions/2558695/javadoc-reuse-for-overloaded-methods

[the remark this answers is not in the document]
Comment (benedeki) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/utils/SchemaUtils.scala]: We can try it, although I had quite some issues with links in ScalaDoc.

**Pull request merged**